numba/cuda/mathimpl.py

```diff
@@ -76,8 +88,20 @@ def lower_boolean_impl(context, builder, sig, args):
 
 def impl_unary(key, ty, libfunc):
     def lower_unary_impl(context, builder, sig, args):
-        libfunc_impl = context.get_function(libfunc, typing.signature(ty, ty))
-        return libfunc_impl(builder, args)
+        if ty == float32 and context.fastmath is True:
+            fast_replacement = unarys_fastmath.get(libfunc.__name__)
+            if fast_replacement is None:
+                libfunc_impl = context.get_function(libfunc,
+                                                    typing.signature(ty, ty))
+            else:
+                new_libfunc = getattr(libdevice, fast_replacement)
+                libfunc_impl = context.get_function(new_libfunc,
+                                                    typing.signature(ty, ty))
+            return libfunc_impl(builder, args)
+        else:
+            libfunc_impl = context.get_function(libfunc,
+                                                typing.signature(ty, ty))
```

Comment:
Think the 3x `libfunc_impl = context.get_function...` can be pulled out as all paths use it, then the branch is just on 32bit float with fastmath and switching the library `libfunc = getattr(libdevice, fast_replacement)`, something like the (entirely untested!) code below perhaps, what do you think?
```python
if ty == float32 and context.fastmath:
    fast_replacement = unarys_fastmath.get(libfunc.__name__)
    if fast_replacement is not None:
        libfunc = getattr(libdevice, fast_replacement)

libfunc_impl = context.get_function(libfunc,
                                    typing.signature(ty, ty))
```